### PR TITLE
Make @asset and @multi_asset return asset definitions instead of ops

### PR DESCRIPTION
**Pull request by sryza**
This should not affect behavior.

A couple consequences:
* Users can no longer take the definitions produced by `@asset` and `@multi_asset` and build jobs out of them using `@graph` / `@job`.  This is a good thing, because in theory they could wire up the graph dependencies in a way that contradicts the asset dependencies, which would cause all kinds of zaniness.
* We'll have a place for the asset partition definitions to live.

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/decorators/op.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 from functools import update_wrapper
-from typing import Any, Callable, Dict, List, Optional, Sequence, Set, Union
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Sequence, Set, Union
```
Comment (sryza) [inline python_modules/dagster/dagster/core/definitions/decorators/op.py:2]: Oops, I'm going to move these changes that don't have anything to do with asset defs to a separate PR.

File: python_modules/dagster/dagster/core/asset_defs/assets_job.py [hunk 2]
```diff
@@ -25,14 +23,15 @@
...
+from .asset import MultiAssetDefinition
 from .foreign_asset import ForeignAsset
 
 
 @experimental
 def build_assets_job(
     name: str,
-    assets: List[OpDefinition],
-    source_assets: Optional[Sequence[Union[ForeignAsset, OpDefinition]]] = None,
+    assets: List[Union[MultiAssetDefinition]],
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/asset_defs/assets_job.py:33]: nit: no need for union

File: python_modules/dagster/dagster/core/asset_defs/decorators.py [hunk 3]
```diff
@@ -67,7 +67,7 @@ def my_asset(my_upstream_asset: int) -> int:
     if callable(name):
         return _Asset()(name)
 
-    def inner(fn: Callable[..., Any]) -> OpDefinition:
+    def inner(fn: Callable[..., Any]) -> MultiAssetDefinition:
```
Comment (OwenKephart) [inline python_modules/dagster/dagster/core/asset_defs/decorators.py:70]: Very much a nitpick, but I find it kinda unaesthetic that `@asset` and `@multi_asset` both return a `MultiAssetDefinition`, because the existence of a `MultiAssetDefinition` implies the existence of an `AssetDefinition`. I get that having just a single asset is just a special case of having multiple assets (and so it's even worse to have some AssetDefinition that subclasses MultiAssetDefinition), but could we instead call `MultiAssetDefinition` something like `AssetsDefinition`?

It feels a little more natural to me that `AssetsDefinition` might define one or more assets than `MultiAssetDefinition` having that behavior (which seems more like it's guaranteed to define more than one).

Comment (sryza) [reply python_modules/dagster/dagster/core/asset_defs/decorators.py:70]: Hmm yeah, I toyed with the name `AssetsDefinition` a little bit. One concern is that if we do choose to introduce `AssetDefinition`, distinguishing between the two might be annoying.  We can cross the bridge when we come to it though - I'll make the change you're suggesting.

**Review by OwenKephart — APPROVED** [review]

File: python_modules/dagster/dagster/core/asset_defs/asset.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from typing import Mapping
+
+from dagster.core.definitions import OpDefinition
+from dagster.core.definitions.events import AssetKey
+
+
+class AssetsDefinition:
```
Comment (prha) [inline python_modules/dagster/dagster/core/asset_defs/asset.py:7]: nit, what's the naming choice here?  why not `AssetDefinition`?

Comment (sryza) [reply python_modules/dagster/dagster/core/asset_defs/asset.py:7]: Each one of these can include multiple assets that are computed from the same op and same dependencies
